src/wrapper.rs

```diff
@@ -0,0 +1,235 @@
+            .is_instance(py.get_type::<PyString>())?
+        {
+            let address: &PyString = server_address_attr.downcast(py)?;
+            format!("input.forward_pipe('{address}')")
+        } else {
+            return Err(py_err(format!(
+                "Unhandled server_address type: {address_type}"
+            )));
+        };
+    }
+
+    loop_wrap(
+        py,
+        tunnel,
+        &format!(
+            r###"
+    if input is None:
+        session = await NgrokSessionBuilder().authtoken_from_env().connect()
+        input = await session.http_endpoint().listen()
+    {forward}
+    return input
+    "###
+        ),
+    )
+}
+
+/// Set the WERKZEUG_SERVER_FD environment variable with a file descriptor from a default HTTP tunnel.
+/// Also sets WERKZEUG_RUN_MAIN to "true" to engage the use of WERKZEUG_SERVER_FD.
+///
+/// Returns the created tunnel if no async loop is running, otherwise returns a Task to await with a tunnel result.
+#[pyfunction]
+pub fn werkzeug_develop(py: Python, tunnel: Option<Py<PyAny>>) -> PyResult<Py<PyAny>> {
+    loop_wrap(
+        py,
+        tunnel,
+        r###"
+    if input is None:
+        session = await NgrokSessionBuilder().authtoken_from_env().connect()
+        input = await session.http_endpoint().listen()
+    os.environ["WERKZEUG_SERVER_FD"] = str(input.fd)
+    os.environ["WERKZEUG_RUN_MAIN"] = "true"
+    return input
+    "###,
+    )
+}
+
+/// Python wrapper to call the passed in work in an async context whether or not an async loop is running.
+fn loop_wrap(py: Python, input: Option<Py<PyAny>>, work: &str) -> PyResult<Py<PyAny>> {
+    let code = format!(
+        r###"
+import asyncio
+from ngrok import NgrokSessionBuilder
+import os
+
+async def wrap(input=None):
+{work}
+
+def run(input=None):
+    try:
+        running_loop = asyncio.get_running_loop()
+        return running_loop.create_task(wrap(input))
+    except RuntimeError:
+        pass
+
+    tunnel = asyncio.run(wrap(input))
+    return tunnel
+    "###
+    );
+
+    call_code(py, input, code.as_str())
+}
+
+/// Call the given code, returning the required 'retval' attribute from it.
+fn call_code(py: Python, input: Option<Py<PyAny>>, code: &str) -> PyResult<Py<PyAny>> {
+    let run = PyModule::from_code(py, code, "", "")?.getattr("run")?;
+
+    let res = match input {
+        Some(input) => {
+            let args = PyTuple::new(py, &[input]);
+            run.call1(args)?
+        }
+        None => run.call0()?,
+    };
+
+    Ok(res.into())
+}
+
+/// Create and bind a python localhost TCP socket.
+pub(crate) fn bound_default_tcp_socket(py: Python) -> PyResult<Py<PyAny>> {
+    let socket = PyModule::import(py, intern!(py, "socket"))?;
+    let sock_func = socket.getattr(intern!(py, "socket"))?;
+    let obj = sock_func.call0()?;
+    let bind = obj.getattr(intern!(py, "bind"))?;
+    let host: &PyAny = PyString::new(py, "localhost");
+    let port: &PyAny = 0u8.into_py(py).into_ref(py);
```

Comment:
I always forget that you can bind a random port with 0!

Reply:
Has been super useful for these convenience wrappers here and on js side 📈